cmd/diag/ui/ui.go

```diff
@@ -0,0 +1,157 @@
+package ui
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"os/exec"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/ledgerwatch/erigonwatch"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/chi/v5/middleware"
+	"github.com/go-chi/cors"
+	"github.com/ledgerwatch/erigon/cmd/diag/flags"
+	"github.com/urfave/cli/v2"
+)
+
+var (
+	UIPortFlag = cli.IntFlag{
+		Name:     "ui.port",
+		Aliases:  []string{"up"},
+		Usage:    "UI port to listen on",
+		Required: false,
+		Value:    8080,
+	}
+)
+
+var Command = cli.Command{
+	Name:      "ui",
+	Action:    runUI,
+	Aliases:   []string{"u"},
+	Usage:     "run local ui",
+	ArgsUsage: "",
+	Flags: []cli.Flag{
+		&flags.DebugURLFlag,
+		&UIPortFlag,
+	},
+	Description: ``,
+}
+
+func runUI(cli *cli.Context) error {
+	supportedSubpaths := []string{
+		"sentry-network",
+		"sentinel-network",
+		"downloader",
+		"logs",
+		"chain",
+		"data",
+		"debug",
+		"testing",
+		"performance",
+		"documentation",
+		"issues",
+		"admin",
+	}
+
+	listenAddr := "127.0.0.1"
+	listenPort := cli.Int(UIPortFlag.Name)
+
+	assets, _ := erigonwatch.UIFiles()
+	fs := http.FileServer(http.FS(assets))
+
+	r := chi.NewRouter()
+	r.Use(middleware.Logger)
+	r.Use(middleware.Recoverer)
+	r.Use(middleware.RouteHeaders().
+		Route("Origin", "*", cors.Handler(cors.Options{
+			AllowedOrigins:   []string{"*"},
+			AllowedMethods:   []string{"GET", "POST", "PUT", "DELETE", "OPTIONS"},
+			AllowedHeaders:   []string{"Accept", "Content-Type", "session-id"},
+			AllowCredentials: false, // <----------<<< do not allow credentials
+		})).
+		Handler)
+
+	r.Mount("/", fs)
+
+	for _, subpath := range supportedSubpaths {
+		addhandler(r, "/"+subpath, fs)
+	}
+
+	// Use the file system to serve static files
+	url := "http://" + cli.String(flags.DebugURLFlag.Name)
+	addr := DiagAddress{
+		Address: url,
+	}
+
+	//r.Get("/diagaddr", writeDiagAdderss(addr))
+	r.Handle("/data", http.StripPrefix("/data", fs))
+
+	r.HandleFunc("/diagaddr", func(w http.ResponseWriter, r *http.Request) {
+		writeDiagAdderss(w, addr)
+	})
+
+	srv := &http.Server{
+		Addr:              fmt.Sprintf("%s:%d", listenAddr, listenPort),
+		Handler:           r,
+		MaxHeaderBytes:    1 << 20,
+		ReadHeaderTimeout: 1 * time.Minute,
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done() // Signal that the goroutine has completed
+		err := srv.ListenAndServe()
+
+		if err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	open(fmt.Sprintf("http://%s:%d", listenAddr, listenPort))
```

Comment:
In my opinion, it is better to provide a log line with the URL rather than attempting to open the browser. There may be cases where the UI is run on a headless machine.

Reply:
User shouldn't run UI command to headless machine as it makes nonsense but it is the good point that is can accidentally happen. I'll fix it.